File: autoimageslider/src/main/java/com/smarteist/autoimageslider/InfiniteAdapter/InfinitePagerAdapter.java
package com.smarteist.autoimageslider.InfiniteAdapter;

import android.database.DataSetObserver;
import android.os.Parcelable;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.viewpager.widget.PagerAdapter;

import com.smarteist.autoimageslider.SliderViewAdapter;


/**
 * Its just a wrapper adapter class for providing infinite behavior
 * for slider.
 */
public class InfinitePagerAdapter extends PagerAdapter {

    // Warning: it should be an even number.
    public static final int INFINITE_SCROLL_LIMIT = 32400;
    private static final String TAG = "InfinitePagerAdapter";
    private SliderViewAdapter adapter;

    public InfinitePagerAdapter(SliderViewAdapter adapter) {
        this.adapter = adapter;
    }

    public PagerAdapter getRealAdapter() {
        return this.adapter;
    }

    @Override
    public int getCount() {
        if (getRealCount() < 1) {
            return 0;
        }
        // warning: infinite scroller actually is not infinite!
        // very big number will be cause memory problems.
        return getRealCount() * INFINITE_SCROLL_LIMIT;
    }

    /**
     * @return the {@link #getCount()} result of the wrapped adapter
     */
    public int getRealCount() {
        try {
            return getRealAdapter().getCount();
        } catch (Exception e) {
            return 0;
        }
    }


    /**
     * @param item real position of item
     * @return virtual mid point
     */
    public int getMiddlePosition(int item) {
        int adapterCount = getRealCount() - 1;
        adapterCount = adapterCount == 0 ? 1 : adapterCount;
        int midpoint = adapterCount * (InfinitePagerAdapter.INFINITE_SCROLL_LIMIT / 2);
        return item + midpoint;
    }

    @NonNull
    @Override
    public Object instantiateItem(ViewGroup container, int virtualPosition) {
        // prevent division by zer
        if (getRealCount() < 1) {
            return adapter.instantiateItem(container, 0);
        }
        //Log.i(TAG, "instantiateItem: real virtualPosition: " + virtualPosition);
        //Log.i(TAG, "instantiateItem: virtual virtualPosition: " + virtualPosition);

        // only expose virtual virtualPosition to the inner adapter
        return adapter.instantiateItem(container, getRealPosition(virtualPosition));
    }

    @Override
    public void destroyItem(ViewGroup container, int virtualPosition, Object object) {
        // prevent division by zero
        if (getRealCount() < 1) {
            adapter.destroyItem(container, 0, object);
            return;
        }
        //Log.i(TAG, "destroyItem: real position: " + position);
        //Log.i(TAG, "destroyItem: virtual position: " + virtualPosition);

        // only expose virtual position to the inner adapter
        adapter.destroyItem(container, getRealPosition(virtualPosition), object);
    }

    @Override
    public void startUpdate(ViewGroup container) {
        adapter.startUpdate(container);
    }

    /*
     * Delegate rest of methods directly to the inner adapter.
     */
    @Override
    public void finishUpdate(ViewGroup container) {
        adapter.finishUpdate(container);
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return adapter.isViewFromObject(view, object);
    }

    @Override
    public void restoreState(Parcelable bundle, ClassLoader classLoader) {
        adapter.restoreState(bundle, classLoader);
    }

    @Override
    public Parcelable saveState() {
        return adapter.saveState();
    }

    @Override
    public CharSequence getPageTitle(int virtualPosition) {
        return adapter.getPageTitle(getRealPosition(virtualPosition));
    }

    @Override
    public float getPageWidth(int position) {
        return adapter.getPageWidth(position);
    }

    @Override
    public void setPrimaryItem(ViewGroup container, int position, Object object) {
        adapter.setPrimaryItem(container, position, object);
    }

    @Override
    public void unregisterDataSetObserver(DataSetObserver observer) {
        adapter.unregisterDataSetObserver(observer);
    }

    @Override
    public void registerDataSetObserver(DataSetObserver observer) {
        adapter.registerDataSetObserver(observer);
    }

    @Override
    public int getItemPosition(Object object) {
        return adapter.getItemPosition(object);
    }

    public int getRealPosition(int virtualPosition) {
        if (getRealCount() > 0) {
            return virtualPosition % getRealCount();
        }
        return 0;
    }
}

File: autoimageslider/src/main/java/com/smarteist/autoimageslider/InfiniteAdapter/InfinitePagerAdapter.kt
package com.smarteist.autoimageslider.InfiniteAdapter

import android.database.DataSetObserver
import android.os.Parcelable
import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import com.smarteist.autoimageslider.SliderViewAdapter

/**
 * Wraps a [SliderViewAdapter] to provide “infinite” scrolling behavior.
 * Internally, it multiplies the real item count by [INFINITE_SCROLL_FACTOR] to avoid
 * reaching the natural ends. All positions are mapped back to the real adapter via modulo.
 */
class InfinitePagerAdapter(
    private val wrappedAdapter: SliderViewAdapter<*>
) : PagerAdapter() {

    /**
     * If the wrapped adapter has no items, we return 0.
     * Otherwise we pretend there are (realCount × INFINITE_SCROLL_FACTOR) pages.
     */
    override fun getCount(): Int {
        val realCount = wrappedAdapter.count
        return if (realCount <= 0) 0 else realCount * INFINITE_SCROLL_FACTOR
    }

    /**
     * Return the actual number of items in the wrapped adapter.
     */
    val realCount: Int
        get() = wrappedAdapter.count.coerceAtLeast(0)

    /**
     * Given a “real” item position, return a virtual midpoint that lands on that item.
     * Useful for starting the pager near the middle of the infinite range.
     */
    fun getMiddlePosition(realPosition: Int): Int {
        val count = realCount
        if (count <= 0) return 0
        val midpoint = (count * (INFINITE_SCROLL_FACTOR / 2))
        return midpoint + realPosition.coerceIn(0, count - 1)
    }

    override fun instantiateItem(container: ViewGroup, virtualPosition: Int): Any {
        val realPos = getRealPosition(virtualPosition)
        return wrappedAdapter.instantiateItem(container, realPos)
    }

    override fun destroyItem(container: ViewGroup, virtualPosition: Int, `object`: Any) {
        val realPos = getRealPosition(virtualPosition)
        wrappedAdapter.destroyItem(container, realPos, `object`)
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return wrappedAdapter.isViewFromObject(view, `object`)
    }

    override fun startUpdate(container: ViewGroup) {
        wrappedAdapter.startUpdate(container)
    }

    override fun finishUpdate(container: ViewGroup) {
        wrappedAdapter.finishUpdate(container)
    }

    override fun saveState(): Parcelable? {
        return wrappedAdapter.saveState()
    }

    override fun restoreState(state: Parcelable?, loader: ClassLoader?) {
        wrappedAdapter.restoreState(state, loader)
    }

    override fun getPageTitle(virtualPosition: Int): CharSequence? {
        return wrappedAdapter.getPageTitle(getRealPosition(virtualPosition))
    }

    override fun getPageWidth(virtualPosition: Int): Float {
        return wrappedAdapter.getPageWidth(getRealPosition(virtualPosition))
    }

    override fun setPrimaryItem(container: ViewGroup, virtualPosition: Int, `object`: Any) {
        wrappedAdapter.setPrimaryItem(container, getRealPosition(virtualPosition), `object`)
    }

    override fun registerDataSetObserver(observer: DataSetObserver) {
        wrappedAdapter.registerDataSetObserver(observer)
    }

    override fun unregisterDataSetObserver(observer: DataSetObserver) {
        wrappedAdapter.unregisterDataSetObserver(observer)
    }

    override fun getItemPosition(`object`: Any): Int {
        return wrappedAdapter.getItemPosition(`object`)
    }

    /**
     * Map any “virtual” position to the equivalent real adapter position.
     */
    fun getRealPosition(virtualPosition: Int): Int {
        val count = realCount
        return if (count > 0) virtualPosition % count else 0
    }

    companion object {
        /**
         * Must be even; determines how many “cycles” we allow before reaching the end.
         *
         * For example, if realCount = 5 and INFINITE_SCROLL_FACTOR = 10, getCount() = 50.
         * The midpoint is at 25, so you can scroll 25 items in either direction before hitting bounds.
         */
        const val INFINITE_SCROLL_FACTOR = 32_400
    }
}
